src/stash/src/transaction.rs

```diff
@@ -0,0 +1,641 @@
+                Result::<_, StashError>::Ok((*collection_id, since))
+            });
+        }
+        let sinces = HashMap::from_iter(try_join_all(futures).await?);
+        Ok(sinces)
+    }
+
+    /// Iterates over a collection.
+    #[tracing::instrument(level = "debug", skip_all)]
+    pub async fn iter<K, V>(
+        &self,
+        collection: StashCollection<K, V>,
+    ) -> Result<Vec<((K, V), Timestamp, Diff)>, StashError>
+    where
+        K: Data,
+        V: Data,
+    {
+        let since = match self.since(collection.id).await?.into_option() {
+            Some(since) => since,
+            None => {
+                return Err(StashError::from(
+                    "cannot iterate collection with empty since frontier",
+                ));
+            }
+        };
+        let rows = self
+            .client
+            .query(self.stmts.iter(), &[&collection.id])
+            .await?
+            .into_iter()
+            .map(|row| {
+                let key: Value = row.try_get("key")?;
+                let value: Value = row.try_get("value")?;
+                let time = row.try_get("time")?;
+                let diff: Diff = row.try_get("diff")?;
+                Ok::<_, StashError>(((key, value), cmp::max(time, since), diff))
+            })
+            // The collect here isn't needed, we just want the short circuit return
+            // behavior of ?. Is there a way to achieve that without allocating a Vec?
+            .collect::<Result<Vec<_>, _>>()?;
+        let rows = consolidate_updates_kv(rows).collect();
+        Ok(rows)
+    }
+
+    /// Iterates over the values of a key.
+    #[tracing::instrument(level = "debug", skip_all)]
+    pub async fn iter_key<K, V>(
+        &self,
+        collection: StashCollection<K, V>,
+        key: &K,
+    ) -> Result<Vec<(V, Timestamp, Diff)>, StashError>
+    where
+        K: Data,
+        V: Data,
+    {
+        let key = serde_json::to_vec(key).expect("must serialize");
+        let key: Value = serde_json::from_slice(&key)?;
+        let (since, rows) = future::try_join(
+            self.since(collection.id),
+            self.client
+                .query(self.stmts.iter_key(), &[&collection.id, &key])
+                .map_err(|err| err.into()),
+        )
+        .await?;
+        let since = match since.into_option() {
+            Some(since) => since,
+            None => {
+                return Err(StashError::from(
+                    "cannot iterate collection with empty since frontier",
+                ));
+            }
+        };
+        let mut rows = rows
+            .into_iter()
+            .map(|row| {
+                let value: Value = row.try_get("value")?;
+                let value: V = serde_json::from_value(value)?;
+                let time = row.try_get("time")?;
+                let diff = row.try_get("diff")?;
+                Ok::<_, StashError>((value, cmp::max(time, since), diff))
```

Comment:
It moves up any timestamps < since to since so that the consolidate below can eliminate things that got deleted but haven't been consolidated away in the underlying cockroach database yet.